man/bytes.1
." vim: set syn=nroff
.TH bytes 1 "May 2019" "bytes v1.5.0"

.SH NAME
bytes - calculate the disk space consumed by files within a directory.

.SH SYNOPSIS
bytes [-ahnrv]
      [-p \fI\,precision\fR]
      [-u \fI\,unit\fR]
      [\fI\,directory\fR]

.SH DESCRIPTION
\fI\,bytes\fR calculates the disk space consumed by files with a directory. \fI\,bytes\fR does not support IEC.

To calculate the total number of bytes consumed by all visible (i.e. non-dot) files within the current directory excluding subdirectories, simply enter

.RS
bytes
.RE

into your terminal.
You can also specify a different directory to calculate the file-size total for:

.RS
bytes \fI\,path/to/directory\fR
.RE

If you want to display the result in a different unit, e.g. kilobytes or megabytes, use the \fI\,-u\fR option to specify the desired unit.

If you want to include hidden files, i.e. dot-files, use the \fI\,-a\fR option.

If you want to include subdirectories, use the \fI\,-r\fR option.

See the \fI\,OPTIONS\fR section for a full list of available options.

.SH OPTIONS
-a, --all
.RS
include dot-files a.k.a. hidden files.
.RE

-h, --help
.RS
display usage information.
.RE

-n, --newline
.RS
do not print a trailing newline character.
.RE

-p \fI\,precision\fR, --precision \fI\,precision\fR
.RS
specify the desired floating point precision. Default: 3.
.RE

-r, --recursive
.RS
if set, \fI\,bytes\fR recursively traverses subdirectories.
.RE

-u \fI\,unit\fR, --unit \fI\,unit\fR
.RS
specify the unit in which to display disk space consumption. Supported units are:
.RS
.IP \fI\,b\fR .2i
bytes (default)
.IP \fI\,kb\fR
kilobytes (1024 bytes)
.IP \fI\,mb\fR
megabytes (1014 kilobytes)
.IP \fI\,gb\fR
gigabytes (1024 megabytes)
.IP \fI\,tb\fR
terabytes (1024 gigabytes)
.RE
.RE

-v, --version
.RS
display version information.
.RE

.SH ARGUMENTS
\fI\,directory\fR
.RS
specify the directory for which to calculate the consumed disk space. Defaults to the current directory.

.SH BUGS
Report bugs to <https://gihub.com/haensl/bytes/issues>

.SH AUTHOR
Written by Hans-Peter Dietz <h.p.dietz@gmail.com>



CHANGELOG.md
## 1.5.1
* Remove obsolete file.

## 1.5.0
* [#10: Add Makefile to install.](https://github.com/haensl/bytes/issues)
* Add `[-h | --help]` option.
* Add manpage.
* Fix precision option.
* Add tests.
* Add CI.

## 1.4.0
* Add `[-r | --recursive]` option

## 1.3.0
* Add `[-v | --version]` option

## 1.2.1
* Check for existence of `[bashrc | bash_profile].local` in install script

## 1.2.0
* Add `[-p | --precision]` option to specify floating point precision of output

## 1.1.1
* Distinguish between .bashrc (Linux) and .bash_profile (OS X) in installation script

## 1.1.0
* Add `[-n | --newline]` option to omit output of trailing newline character

## 1.0.0
* bytes initial release


bytes.sh
#!/usr/bin/env bash

# bytes
# Sum the number of bytes in a directory listing
# Written by HP Dietz | h.p.dietz@gmail.com


set -e
readonly VERSION=1.5.1
readonly ARGS="$@"

usage() {
  echo "Usage: bytes [-u | --unit <b | kb | mb | gb | tb>] [-a | --all] [-n | --newline] [-p | --precision <precision>] [-v | --version] [directory]" >&2
}

version() {
  echo "bytes v$VERSION" >&2
}


printDirSize() {
  local bytes=$1
  output=$(printf "%.${PRECISION}f" $(echo -e "scale=$PRECISION \n$bytes/$UNIT_FACTOR \nquit" | bc))
  if $PRINT_NEWLINE ; then
    echo "$output"
  else
    echo -n "$output"
  fi
}

getUnitFactor() {
  local unit=$1
  local outVar=$2
  case "$unit" in
    b)
      eval $outVar=1
      ;;
    kb)
      eval $outVar=1024
      ;;
    mb)
      let f=1024**2
      eval $outVar=$f
      ;;
    gb)
      let f=1024**3
      eval $outVar=$f
      ;;
    tb)
      let f=1024**4
      eval $outVar=$f
      ;;
    *)
      echo "Invalid unit: $unit" >&2;
      usage
      exit 1
      ;;
  esac
}

getPrecision() {
  local p=$1
  local outVar=$2

  if ! [[ "$p" =~ ^[0-9]+$ ]]; then
    echo "Invalid precision: $p" >&2
    usage
    exit 1
  fi
  eval $outVar=$p
}

parseArgs() {
  local arg
  for arg 
  do
    case "$arg" in
      # translate --long-options into short options
      --all)        args="${args}-a " ;;
      --help)       args="${args}-h " ;;
      --newline)    args="${args}-n " ;;
      --precision)  args="${args}-p " ;;
      --recursive)  args="${args}-r " ;;
      --unit)       args="${args}-u " ;;
      --version)    args="${args}-v " ;;
      # pass through anything else
      *)            args="${args} ${arg} " ;;
    esac
  done

  eval set -- $args

  local lsOpts=l
  local newline=true
  local precision=3
  local recursive=false
  local unitFactor=1

  while getopts "ahnp:ru:v" OPTION;
  do
    case $OPTION in
      a)
        lsOpts=${lsOpts}a
        ;;
      n)
        printNewline=false
        ;;
      p)
        getPrecision "$OPTARG" precision
        ;;
      r)
        recursive=true
        ;;
      u)
        getUnitFactor "$OPTARG" unitFactor
        ;;
      v)
        version
        exit 0
        ;;
      h)
        usage
        exit 0
        ;;
      *)
        usage
        exit 1
        ;;
    esac
  done
  shift $((OPTIND-1))

  local dir=.
  if [ -n "$1" ]; then 
    dir=$1
    if [ ! -d "$dir" ];
    then
      echo "No such directory: $dir"
      usage
      exit 1
    fi
  fi

  readonly LS_OPTS=$lsOpts
  readonly NEWLINE=$newline
  readonly PRECISION=$precision
  readonly UNIT_FACTOR=$unitFactor
  readonly RECURSIVE=$recursive
  readonly DIR=$dir
}

getBytesInDir() {
  local dir=$1
  local recursive=$2
  local total=0

  for bytes in $(ls -$LS_OPTS $dir | grep "^-" | awk '{ print $5 }')
  do
    let total=$total+$bytes
  done

  if $recursive ; then
    local t=0
    for subDir in $(ls -$LS_OPTS $dir | grep "^d" | awk '{ print $9 }')
    do
      if [ "$subDir" != "." ] && [ "$subDir" != ".." ]; then
        let total=$total+"$(getBytesInDir $dir/$subDir $recursive)"
      fi
    done
  fi

  echo $total
}

main() {
  parseArgs $ARGS
  printDirSize $(getBytesInDir $DIR $RECURSIVE)
}

main

